File: BabyTracker/Models/SyncTypes.swift
import Foundation

/// Every synced model carries this status.
/// The SyncManager uses it to determine what needs pushing to Supabase.
enum SyncStatus: String, Codable {
    case pending    // created/modified locally, not yet sent to server
    case synced     // matches server state
    case failed     // last sync attempt failed (will retry)
    case deleted    // soft-deleted locally, pending remote deletion
}

/// Represents a pending sync operation. Persisted to survive app restarts.
struct SyncOperation: Identifiable, Codable {
    let id: UUID
    let modelType: String       // e.g. "Baby", "FeedEntry"
    let modelID: UUID
    let operation: SyncOperationType
    let payload: Data           // JSON-encoded model snapshot
    var retryCount: Int
    var lastAttemptAt: Date?
    let createdAt: Date

    init(modelType: String, modelID: UUID, operation: SyncOperationType, payload: Data) {
        self.id = UUID()
        self.modelType = modelType
        self.modelID = modelID
        self.operation = operation
        self.payload = payload
        self.retryCount = 0
        self.lastAttemptAt = nil
        self.createdAt = Date()
    }
}

enum SyncOperationType: String, Codable {
    case insert
    case update
    case delete
}

/// Network reachability state used by the sync manager
enum NetworkState {
    case online
    case offline
}

File: flutter_app/lib/models/sync_types.dart
import 'package:uuid/uuid.dart';

enum SyncStatus { pending, synced, failed, deleted }

enum SyncOperationType { insert, update, delete }

class SyncOperation {
  final String id;
  final String modelType;
  final String modelId;
  final SyncOperationType operation;
  final Map<String, dynamic> payload;
  int retryCount;
  DateTime? lastAttemptAt;
  final DateTime createdAt;

  SyncOperation({
    required this.modelType,
    required this.modelId,
    required this.operation,
    this.payload = const {},
  })  : id = const Uuid().v4(),
        retryCount = 0,
        createdAt = DateTime.now();

  Map<String, dynamic> toJson() => {
        'id': id,
        'modelType': modelType,
        'modelId': modelId,
        'operation': operation.name,
        'payload': payload,
        'retryCount': retryCount,
        'lastAttemptAt': lastAttemptAt?.toIso8601String(),
        'createdAt': createdAt.toIso8601String(),
      };

  factory SyncOperation.fromJson(Map<String, dynamic> j) {
    final op = SyncOperation(
      modelType: j['modelType'] as String,
      modelId: j['modelId'] as String,
      operation: SyncOperationType.values.byName(j['operation'] as String),
      payload: (j['payload'] as Map<String, dynamic>?) ?? {},
    );
    op.retryCount = j['retryCount'] as int? ?? 0;
    final lat = j['lastAttemptAt'];
    if (lat != null) op.lastAttemptAt = DateTime.parse(lat as String);
    return op;
  }
}
